[Caleb Cherng Kai Zhe] iP

Ditzchann opened the pull request:
# Angela
> Hello X. I am Angela, your advisor and secretary.

Based on Project Moon's [Lobotomy Corporation](https://store.steampowered.com/app/568220/Lobotomy_Corporation__Monster_Management_Simulation/), Angela is an advanced ~~human~~ AI whose role is to guide you through your tasks just as she did in LC.

With the chat bot, you can enjoy
- the ability to organize your tasks your tasks in one place
- your tasks all saved in your local storage to come back to anytime
- customized dialogue with the head AI of Lobotomy Corp, tailored for any LC fan to enjoy

Features: 
- [x] Saving your tasks into your local storage
- [x] Searching for tasks
- [x] Deleting tasks

Todo (with priority!) 
1. Command class to simplify functions
2. GUI

If you want to contribute, open a pull request to help improve my code! (My poor `processDb` function needs help :sob:)
```
public Task processDb(String line) throws AngelaException {
        Task t;
        String[] lineArr = line.split("\\|\\|"); //no way malicious injection is real
        t = switch (lineArr[0]) { //as we all know nothing bad ever happens when we save data as text
            case "T" -> new ToDoTask(lineArr[2]);
            case "D" -> new DeadlineTask(lineArr[2], lineArr[3]);
            case "E" -> new EventTask(lineArr[2], lineArr[3], lineArr[4]);
            default -> throw new AngelaException("");
        };
        if (lineArr[1].equals("1")) t.doTask();
        return t;
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Ditzchann committed:
Level-0

Added default greet function
Added exit function

Ditzchann committed:
Level-1

Added echo function to repeat user input

Ditzchann committed:
Level-2

Added Processor.java to assist in text processing
Added functionality to store inputs as a list and output them accumulatively

Ditzchann committed:
Created Task.java for task representation

Created initialization for Task class
Added toString() method for Task class
Added ability to set tasks as done

Ditzchann committed:
Added ability to do and undo tasks to Processor.java

Updated Task.java to be able to mark tasks as done
Added markDone() and markUndone() to Processor.java

Ditzchann committed:
Added DeadlineTask.java and Task.java

Ditzchann committed:
Added EventTask.java

Added EventTask.java
Fixed some formatting errors

Ditzchann committed:
Added trimInput function to help process functions

Ditzchann committed:
Added support for multiple tasks types

Added processing for DeadlineTask, TodoTask and EventTask

RT0118 commented on src/main/java/DeadlineTask.java:
```diff
@@ -0,0 +1,12 @@
+
+	@Override
+	public String toString() {
+		return "[D]" + super.toString()  + " (by: " + by + ")";
```
I like the use of newlines to keep char count below 120

RT0118 commented on src/main/java/Processor.java:
```diff
@@ -0,0 +1,136 @@
+import java.util.regex.Pattern;
+
+public class Processor {
+	private List<Task> store = new ArrayList<>();
```
```suggestion
	private List<Task> store = new ArrayList<>();
```
Should this be named more specific?

RT0118 commented on src/main/java/Angela.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Scanner;
+import java.util.List;
```
```suggestion
import java.util.List;
import java.util.Scanner;
```
Perhaps your imports could be ordered lexicographically?

RT0118 reviewed (commented):
Other than a few nits, LGTM!

nikingoda commented on src/main/java/Processor.java:
```diff
@@ -0,0 +1,136 @@
+		String input;
+		do {
+			input = textIn.nextLine();
+			if (input.equalsIgnoreCase("list")) {
```
Maybe you can use switch-case style to improve readability and maintainability in future

nikingoda commented on src/main/java/Processor.java:
```diff
@@ -0,0 +1,136 @@
+		output(out.trim(), true);
+	}
+
+	public void addTask(String input) {
```
maybe addTask(Task task) be better instead of 3 versions of addTask() for Event, Deadline and Todo, to avoid misusing

nikingoda commented on src/main/java/Processor.java:
```diff
@@ -0,0 +1,136 @@
+			input = textIn.nextLine();
+			if (input.equalsIgnoreCase("list")) {
+				printList();
+			} else if (input.length() >= 4 && input.substring(0, 4).equalsIgnoreCase("mark")) { //probably move this in the future
```
you may use split(" ") to make an array String[] for convenience

nikingoda reviewed (commented):
Overall, your code is easy to read and understand. However, there are still spaces to improve. 
Good job!

SomneelSaha2042 commented on src/main/java/Angela.java:
The order of imports could be lexicographical?

SomneelSaha2042 commented on src/main/java/Processor.java:
A switch case would be easier to follow in term of logic inside the start() function.
We can use .split() to get the different parts of the input?
Then once we get the command we continue to process based off of  expected format.
Since the list of tasks is of type Task we can modify addTask to have signature addTask(Task t) as then we can add all 3 types without having to write 3 different methods.

SomneelSaha2042 commented on src/main/java/Processor.java:
We can use the java string package to more efficiently process input?
The .trim() and .split() method could remove the need for the trimInput functions

SomneelSaha2042 commented on src/main/java/Task.java:
We can make this class abstract, as we wont be instantiating any objects of type Task and we can prevent misuse of this class and enforce usage of the actual 3 types we expect.

SomneelSaha2042 reviewed (commented):
LGTM!
There a are a few minor refactorings that are non essential.

Ditzchann committed:
Add Output.java class

Output messages can be concatenated into a class and called using its static methods to prevent code duplication

Ditzchann committed:
Update Processor.java

Cleaned up command processing by adding new function to split command input
Used switch-case instead of if-else to polish code

Ditzchann committed:
Use Output.java class in Processor

More abstraction in the Processor class

Ditzchann committed:
Edit text-ui-test

Add testcases to input.txt and EXPECTED.TXT
Modify runtest.bat

Ditzchann committed:
Add AngelaException.java

Add several Exceptions to handle cases of invalid input

Ditzchann committed:
Add delete() method to Processor.java

Add ability to delete tasks that are made

Ditzchann committed:
Add getArguments method

Add getArguments method to ensure correct argument flags are passed in

Ditzchann committed:
Add Storage.java

Storage.java manages the storage of the tasks into local files.
Edit Processor.java to integrate changes from Storage.java

Ditzchann committed:
Merge pull request #2 from Ditzchann/branch-Level-7

Branch level 7

Ditzchann committed:
Add ParseTime() method to Task.java

Ditzchann committed:
Add support for dates in DD-MM-YYYY format

Ditzchann committed:
Merge pull request #3 from Ditzchann/branch-Level-8

Branch level 8

Ditzchann committed:
Add Command.java class

Add Command.java class to act as data structure for commands

Ditzchann committed:
Reorganise code in Processor into other classes

Add Parser.java and TaskManager.java to handle command parsing and editing the task list

Ditzchann committed:
Merge pull request #4 from Ditzchann/branch-A-MoreOOP

Branch a more oop

Ditzchann committed:
Package java class into packages

Ditzchann committed:
Merge pull request #5 from Ditzchann/A-Packages

Package java class into packages

Ditzchann committed:
Merge pull request #6 from Ditzchann/add-gradle-support

Add gradle support

Ditzchann committed:
Update build.gradle to reflect correct class

Ditzchann committed:
Add test for Parser.java and Command.java

Ditzchann committed:
Update build.gradle for making JAR file

Ditzchann committed:
Add JavaDoc

Ditzchann committed:
Fix coding standard issues

Ditzchann committed:
Add method to find tasks by keyword

Ditzchann committed:
Merge pull request #7 from Ditzchann/branch-Level-9

Add method to find tasks by keyword

Ditzchann committed:
Merge branch 'master' into A-CodingStandard

Ditzchann committed:
Merge pull request #8 from Ditzchann/A-CodingStandard

Fix coding standard issues

Ditzchann committed:
Merge branch 'master' into A-JavaDoc

Ditzchann committed:
Merge pull request #9 from Ditzchann/A-JavaDoc

Add JavaDoc

Ditzchann committed:
Update JavaDoc in TaskManager.java

Ditzchann committed:
Fix checkstyle errors

Ditzchann committed:
Remove Processor.java

Ditzchann committed:
Add basic support for GUI

Ditzchann committed:
Rename methods

Ditzchann committed:
Merge pull request #10 from Ditzchann/branch-level-10

Branch level 10

Ditzchann committed:
Add assertions to Angela.java and Storage.java

Ditzchann committed:
Add comments and fix code quality issues

Ditzchann committed:
Update Angela.java to handle undo case

Add lastFunction local variable to Angela.java

Let's create a variable that stores the last inputted command that
was called by the user for use in the undo function.

This allows for the undo method to recall the last function used
and act accordingly in the future.

Add new undo() method to Angela.java

The method will recall the last function, calling an undo method in
the manager if the last function made changes to the database.

In the event of a bad undo (undoing a list, find command), the
method will give an error message.

Ditzchann committed:
Add undo functionality

Task class modifications

Task class and its subclasses are not effectively final.

This caused problems when archiving the previous iteration
of the task list.

Let's change do/undo task methods to return a new copy
of the task, allowing for easier archiving for the tasks.

Add support for archiving task list in TaskManager

Lets let TaskManager save a copy of the previous iteration
of tasks before any changes to the tasklist such that
it can be easily retrieved to undo the change.

Ditzchann committed:
Merge pull request #11 from Ditzchann/branch-A-Assertions

Add assertions to Angela.java and Storage.java

Ditzchann committed:
Merge branch 'master' into branch-A-CodeQuality

Ditzchann committed:
Merge pull request #12 from Ditzchann/branch-A-CodeQuality

Improve code quality

Ditzchann committed:
Merge branch 'master' into branch-BCD-Extension

Ditzchann committed:
Merge branch 'master' into branch-BCD-Extension

Ditzchann committed:
Merge pull request #13 from Ditzchann/branch-BCD-Extension

Add support for undoing commands

Ditzchann committed:
Fix minor bug

Event Tasks were saved in the wrong format

Ditzchann committed:
Add attribution

Ditzchann committed:
Add user guide and Ui screenshot

Ditzchann committed:
Update README.md
